### PR TITLE
Update app to target Android 10 

### PR DESCRIPTION
**Pull request by ShivamPokhriyal**

### CODE REVIEW

File: app/src/org/commcare/CommCareApplication.java [hunk 1]
```diff
@@ -410,20 +410,27 @@ public void initializeGlobalResources(CommCareApp app) {
     }
 
     @NonNull
+    @Deprecated
     public String getPhoneId() {
```
Comment (ShivamPokhriyal) [inline app/src/org/commcare/CommCareApplication.java]: Not Sure about the use-case of this particular method, but it won't work with android 10.
Android best practices suggest [See 3rd option](https://developer.android.com/training/articles/user-data-ids#best-practices-android-identifiers) using either a random UUID(we'll have to save it in preferences to make sure it doesn't change) or Firebase Instance ID([read here](https://firebase.google.com/docs/reference/android/com/google/firebase/iid/FirebaseInstanceId)). 

Though both of them have a drawback: if application is reinstalled or user clears app data then the id will be lost and a new id will be generated. 

Let me know, if this is something important and we should use one of the above approaches to get a id.

File: app/AndroidManifest.xml [hunk 2]
```diff
@@ -64,6 +64,7 @@
         android:label="@string/application_name"
         android:hardwareAccelerated="true"
         android:usesCleartextTraffic="true"
+        android:requestLegacyExternalStorage="true"
```
Comment (ShivamPokhriyal) [inline app/AndroidManifest.xml]: Read here for more information :: https://developer.android.com/training/data-storage#scoped-storage
If not enabled, this method throws error https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/models/ODKStorage.java#L22

Comment (ShivamPokhriyal) [inline app/AndroidManifest.xml]: Also, https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/utils/MediaUtil.java#L500 BitmapFactory decodeFile fails if we try and decode the image chosen from gallery.

File: app/src/org/commcare/views/widgets/MediaWidget.java [hunk 3]
```diff
@@ -237,21 +235,12 @@ public void setBinaryData(Object binaryuri) {
             }
         }
 
-        if (newMedia.exists()) {
-            // Add the copy to the content provider
```
Comment (ShivamPokhriyal) [inline app/src/org/commcare/views/widgets/MediaWidget.java]: Though I'm not sure whether we should be adding all the files to the content provider. 
IMO, it makes sense for the video that are recorded using the record button in app, but for videos that are chosen from gallery, we just make a copy of that inside our app's directory so if we add it to content provider there'll be 2 copies of the same video shown to the user in the gallery app.

Comment (ShivamPokhriyal) [inline app/src/org/commcare/views/widgets/MediaWidget.java]: PS: Though, this change will only prevent crashes and won't change the earlier behaviour.

**Comment by ShivamPokhriyal** [comment]
@shubham1g5 @ctsims This is still in progress, but would like to know your opinions on the changes.

File: app/AndroidManifest.xml (hunk 2 above)
Comment (shubham1g5) [reply app/AndroidManifest.xml]: For Android 10 adding this sounds good. Though we should add Soft asserts log  statements to the code paths that we identified that were still accesing code paths. So that we know in future if they are indeed getting hit or not.

File: app/src/org/commcare/CommCareApplication.java (hunk 1 above)
Comment (shubham1g5) [reply app/src/org/commcare/CommCareApplication.java]: So `deviceId` was useful to identify the 2 separate installations of the app. I think using Firebase Instance Id seems like a good workaround here to separate out the installations as we don't have any other id currently that we maintain that's unique per CC installation.

File: app/src/org/commcare/utils/MediaUtil.java [hunk 4]
```diff
@@ -414,7 +414,7 @@ private static int getApproxScaleDownFactor(int newWidth, int originalWidth) {
         double widthScaleDownFactor =  (double)boundingWidth / originalWidth;
         // Choosing the larger of the scale down factors, so that the image still fills the entire
         // container
-        double dominantScaleDownFactor = Math.max(widthScaleDownFactor, heightScaleDownFactor);
+        double dominantScaleDownFactor = Math.min(widthScaleDownFactor, heightScaleDownFactor);
```
Comment (shubham1g5) [inline app/src/org/commcare/utils/MediaUtil.java]: Thanks for looking at it as part of this work. Though if it isn't realted to Android 10 work, would be nice to separate this out in another PR so that it's easy to revert in case of any edge cases.

File: app/src/org/commcare/views/widgets/MediaWidget.java (hunk 3 above)
Comment (shubham1g5) [reply app/src/org/commcare/views/widgets/MediaWidget.java]: Agree that we should not save it to content Providers. @ctsims we did remove it for Images sometime back in [here](https://github.com/dimagi/commcare-android/pull/2110/files), so guessing no objections in removing them for audio and videos as well.

File: app/src/org/commcare/views/widgets/MediaWidget.java [hunk 5]
```diff
@@ -237,21 +235,12 @@ public void setBinaryData(Object binaryuri) {
...
-
-            Uri mediaUri =
-                    getContext().getContentResolver().insert(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, values);
-            String mediaUriString = mediaUri == null ? "null" : mediaUri.toString();
-            Log.i(TAG, "Inserting media returned uri = " + mediaUriString);
+        try {
+            FileUtil.addMediaToGallery(getContext(), newMedia);
             showToast("form.attachment.success");
-        } else {
+        } catch (Exception e) {
```
Comment (shubham1g5) [inline app/src/org/commcare/views/widgets/MediaWidget.java]: We should let the app crash in case of exceptions here if we are going to retain this functionality.

File: app/AndroidManifest.xml (hunk 2 above)
Comment (ctsims) [reply app/AndroidManifest.xml]: We likely need to explicitly request permissions to the filepath beforehand the same way that we now do for things like camera widget capture. I think Shubham reviewed usage of File handles previously, but some of these places where raw paths are used likely didn't get caught.

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

File: app/src/org/commcare/utils/DeviceIdentifier.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+     * Returns a unique(very highly likely) identifier for this device.
+     * This identifier is prefixed with "commcare_"
+     *
+     * This id may change when user re-installs the app or does a factory reset.
+     */
+    public static String getDeviceIdentifier(Context context) {
+        if (uuid == null) {
+            synchronized (DeviceIdentifier.class) {
+                if (uuid == null) {
+                    SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
```
Comment (ShivamPokhriyal) [inline app/src/org/commcare/utils/DeviceIdentifier.java]: Using SharedPreference to store this unique id. We can also choose to create a text file and store UUID to that file. I'm not sure which option is more suitable of the two so using the sharedpreferences approach here.

File: app/src/org/commcare/activities/MultimediaInflaterActivity.java [hunk 7]
```diff
@@ -162,6 +164,7 @@ private void searchForDefault() {
 
         if (Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState()) ||
                 Environment.MEDIA_MOUNTED_READ_ONLY.equals(Environment.getExternalStorageState())) {
+            Logger.log(LogTypes.SOFT_ASSERT, "Access outside scoped storage in searching zip file");
```
Comment (shubham1g5) [inline app/src/org/commcare/activities/MultimediaInflaterActivity.java]: It seems like this path will definitely get hit if someone is using this functionality. So there is no point in adding soft assert here and we will likely need to disable this functionality in Android 11 if there is no other alternate provided in Android 11 for file searching.

File: app/src/org/commcare/recovery/measures/ScanCczTask.java [hunk 8]
```diff
@@ -94,6 +96,7 @@ private void locateAppCcz(File[] files, int depth) {
         if (!StringUtils.isEmpty(lastKnownCczLocation)) {
             filePathsToScan.add(new File(lastKnownCczLocation));
         }
+        Logger.log(LogTypes.SOFT_ASSERT, "Access outside scoped storage in scanning cczs.");
```
Comment (shubham1g5) [inline app/src/org/commcare/recovery/measures/ScanCczTask.java]: We can remove this soft assert as well as this path will definitely get hit for someone using this functionality,

File: app/src/org/commcare/utils/DeviceIdentifier.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    private static final String KEY_DEVICE_IDENTIFIER = "commcare_device_id";
+
+    /**
+     * Returns a unique(very highly likely) identifier for this device.
+     * This identifier is prefixed with "commcare_"
+     *
+     * This id may change when user re-installs the app or does a factory reset.
+     */
+    public static String getDeviceIdentifier(Context context) {
+        if (uuid == null) {
```
Comment (shubham1g5) [inline app/src/org/commcare/utils/DeviceIdentifier.java]: there are 2 nested if here with `uuid == null` check. Is there a particular reason behind that ?

Comment (ShivamPokhriyal) [reply app/src/org/commcare/utils/DeviceIdentifier.java]: Yeah, this pattern is also popular as _double-checked locking idiom_, used to reduce the overhead of acquiring a lock by first testing the locking criterion without actually acquiring the lock.

**Comment by ShivamPokhriyal** [comment]
@damagatchi retest this please

File: app/src/org/commcare/utils/DeviceIdentifier.java (hunk 9 above)
Comment (shubham1g5) [reply app/src/org/commcare/utils/DeviceIdentifier.java]: Interesting!

File: app/src/org/commcare/utils/DeviceIdentifier.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+import android.content.Context;
+import android.content.SharedPreferences;
+import androidx.preference.PreferenceManager;
+import java.util.UUID;
+
+/**
+ * A Utility to get a unique identifier for app's installation.
+ * @author $|-|!˅@M
+ */
+public final class DeviceIdentifier {
```
Comment (shubham1g5) [inline app/src/org/commcare/utils/DeviceIdentifier.java]: @ctsims  Goint to merge this to go ahead with code freeze though can you review this class and see the device identifier makes sense to you.
